File: 백준/Gold/3967. 매직 스타/매직 스타.py
import sys
from collections import deque

ALPHABETS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

input = sys.stdin.readline


def check():
    ops = [
        [0, 2, 5, 7],
        [0, 3, 6, 10],
        [1, 2, 3, 4],
        [1, 5, 8, 11],
        [4, 6, 9, 11],
        [7, 8, 9, 10],
    ]

    for op in ops:
        result = 0
        for i in op:
            result += numbers[i]
        if result != 26:
            return False

    return True


def printAnswer():
    que = deque(numbers)
    for i in range(len(board)):
        for j in range(len(board[0])):
            if board[i][j] != ".":
                board[i][j] = ALPHABETS[que.popleft() - 1]

    for row in board:
        print("".join(row))


def solve(curr):
    if curr == 12:
        if check():
            printAnswer()
            sys.exit()
        return

    if numbers[curr] != 0:
        solve(curr + 1)
    else:
        for i in range(1, 13):
            if not isSelected[i]:
                isSelected[i] = True
                numbers[curr] = i
                solve(curr + 1)
                numbers[curr] = 0
                isSelected[i] = False


board = [list(input().strip()) for _ in range(5)]
numbers = []
for i in range(len(board)):
    for j in range(len(board[0])):
        if board[i][j] == "x":
            numbers.append(0)
        elif board[i][j] != ".":
            numbers.append(ALPHABETS.index(board[i][j]) + 1)


isSelected = [False] * 13
for num in numbers:
    isSelected[num] = True

solve(0)

File: 백준/Gold/3967. 매직 스타/매직 스타.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static char[][] board;
  static int[][] positions;
  static int[] numbers;
  static boolean[] isSelected;

  public static void main(String[] args) throws IOException {
    setUp();

    permutation(0);
    for (int i = 0; i < 12; i++) {
      int y = positions[i][0];
      int x = positions[i][1];
      board[y][x] = (char) ('A' + numbers[i] - 1);
    }

    for (int i = 0; i < 5; i++) {
      for (int j = 0; j < 9; j++) {
        sb.append(board[i][j]);
      }
      sb.append("\n");
    }

    output();
  }

  private static boolean permutation(int curr) {
    if (curr == 12) {
      return check();
    }

    if (numbers[curr] != -1) {
      return permutation(curr + 1);
    }

    for (int i = 1; i < 13; i++) {
      if (!isSelected[i]) {
        isSelected[i] = true;
        numbers[curr] = i;
        if (permutation(curr + 1)) {
          return true;
        }
        numbers[curr] = -1;
        isSelected[i] = false;
      }
    }

    return false;
  }

  private static boolean check() {
    int target = 26;
    int[][] indexes = new int[][]{
        {0, 2, 5, 7},
        {0, 3, 6, 10},
        {1, 2, 3, 4},
        {1, 5, 8, 11},
        {4, 6, 9, 11},
        {7, 8, 9, 10}
    };

    for (int[] index : indexes) {
      int s = 0;
      for (int i : index) {
        s += numbers[i];
      }
      if (s != target) {
        return false;
      }
    }

    return true;
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    board = new char[5][9];
    positions = new int[12][2];
    numbers = new int[12];
    isSelected = new boolean[13];
    int cursor = 0;
    for (int i = 0; i < 5; i++) {
      String line = br.readLine();
      for (int j = 0; j < 9; j++) {
        board[i][j] = line.charAt(j);
        if (board[i][j] != '.') {
          positions[cursor][0] = i;
          positions[cursor][1] = j;
          numbers[cursor] = board[i][j] == 'x' ? -1 : board[i][j] - 'A' + 1;
          cursor++;
        }
      }
    }

    for (int num : numbers) {
      if (num != -1) {
        isSelected[num] = true;
      }
    }
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}